doc/man3/SRP_create_verifier.pod

```diff
@@ -0,0 +1,104 @@
+
+SRP_create_verifier,
+SRP_create_verifier_BN,
+SRP_check_known_gN_param,
+SRP_get_default_gN
+- SRP authentication primitives
+
+=head1 SYNOPSIS
+
+ #include <openssl/srp.h>
+
+ char *SRP_create_verifier_BN(const char *user, const char *pass, BIGNUM **salt,
+                              BIGNUM **verifier, const BIGNUM *N, const BIGNUM *g);
+ char *SRP_create_verifier(const char *user, const char *pass, char **salt,
+                           char **verifier, const char *N, const char *g);
+ 
+ char *SRP_check_known_gN_param(const BIGNUM *g, const BIGNUM *N);
+ SRP_gN *SRP_get_default_gN(const char *id);
+
+=head1 DESCRIPTION
+
+The SRP_create_verifier_BN() function creates a SRP password verifier from
+the supplied parameters as defined in section 2.4 of RFC 5054.
+On successful exit *verifier will point to a newly allocated BIGNUM containing
+the verifier and (if a salt was not provided) *salt will be populated with a
+newly allocated BIGNUM containing a random salt. If *salt is not NULL then
+the provided salt is used instead.
+The caller is responsible for freeing the allocated *salt and *verifier BIGNUMS
+(use L<BN_free(3)>).
+
+The SRP_create_verifier() function is similar to SRP_create_verifier_BN() but
+all numeric parameters are in base64.
+It is possible to pass NULL as B<N> and a SRP group id as B<g> instead to
+load the appropriate gN values. If both B<N> and B<g> are NULL the 8192-bit
+SRP group parameters are used.
+The caller is responsible for freeing the allocated *salt and *verifier char*
+(use L<OPENSSL_free(3)>).
+
+The SRP_check_known_gN_param() function checks that B<g> and B<N> are valid
+SRP group parameters from RFC 5054 appendix A.
+
+The SRP_get_default_gN() function returns the gN parameters for the RFC 5054 B<id>
+SRP group size.
+
+=head1 RETURN VALUES
+
+SRP_create_verifier_BN() returns 1 on success and 0 on failure.
+
+SRP_create_verifier() returns NULL on failure and a non-NULL value on success:
+"*" if B<N> is not NULL, the selected group id otherwise.
+
+SRP_check_known_gN_param() returns the text representation of the group id
+(ie. the prime bit size) or NULL if the arguments are not valid SRP group parameters.
+
+SRP_get_default_gN() returns NULL if B<id> is not a valid group size,
+or the 8192-bit group parameters if B<id> is NULL.
+
+=head1 EXAMPLES
+
+Generate and store a 8192 bit password verifier:
+
+ #include <openssl/bn.h>
+ #include <openssl/srp.h>
+
+ const char *username = "username";
+ const char *password = "password";
+
+ SRP_VBASE *srpData = SRP_VBASE_new(NULL);
+
+ SRP_user_pwd *pwd = (SRP_user_pwd*) OPENSSL_malloc(sizeof(SRP_user_pwd));
+ SRP_gN *gN = SRP_get_default_gN("8192");
+
+ BIGNUM *salt = NULL, *verifier = NULL;
+ SRP_create_verifier_BN(username, password, &salt, &verifier, gN->N, gN->g);
+
+ pwd->id = OPENSSL_strdup(username);
+ pwd->g = gN->g;
+ pwd->N = gN->N;
+ pwd->s = salt;
+ pwd->v = verifier;
+ pwd->info = NULL;
+
+ sk_SRP_user_pwd_push(srpData->users_pwd, pwd);
```

Comment:
There are several such functions in [srp_vfy.c](https://github.com/openssl/openssl/blob/master/crypto/srp/srp_vfy.c#L187) already but they are all marked as static. What should we do about this?

Reply:
I think where they already exist we should make them non-static and public. So: SRP_user_pwd_new, SRP_user_pwd_set_gN, SRP_user_pwd_set_ids, SRP_user_pwd_set_sv_BN.

I would rename all the *set* functions to *set0* to indicate that the memory ownership passes, with the exception of set_ids which I would call set1_ids. I'd also change the setters to free any currently set value first. I'd rename SRP_user_pwd_set_sv_BN to SRP_user_pwd_set0_sv and not expose the existing SRP_user_pwd_set_sv which uses the non-standard base64.

Reply:
SRP_user_pwd_set_gN doesn't really take ownership of its parameters since g and N are not freed by SRP_user_pwd_free. Changing the name might make users think they need to duplicate the output of SRP_get_default_gN before calling this function even though it is not necessary.

I am not sure the _set1_ name would make more sense since there is no duplication involved and freeing the parameters one way or an other makes this structure invalid.

Reply:
Ah - good point. In that case I'd just make it "set" without 0 or 1 and make sure we explain the semantics in the man page.